ui/redux/actions/collections.js

```diff
@@ -333,20 +335,53 @@ export const doFetchItemsInCollection = (options: { collectionId: string, pageSi
   return doFetchItemsInCollections(newOptions, cb);
 };
 
-export const doCollectionEdit = (collectionId: string, params: CollectionEditParams) => async (
+export const doCollectionEdit = (collectionId: string, params: CollectionEditParams) => (
   dispatch: Dispatch,
   getState: GetState
 ) => {
   const state = getState();
-  const collection: Collection = makeSelectCollectionForId(collectionId)(state);
+  const collection: Collection = selectCollectionForId(state, collectionId);
+
+  if (!collection) {
+    return dispatch({
+      type: ACTIONS.COLLECTION_ERROR,
+      data: { message: 'collection does not exist' },
+    });
+  }
+
+  const editedCollection: Collection = selectEditedCollectionForId(state, collectionId);
+  const unpublishedCollection: Collection = selectUnpublishedCollectionForId(state, collectionId);
+  const publishedCollection: Collection = selectPublishedCollectionForId(state, collectionId); // needs to be published only
 
-  if (!collection) return dispatch({ type: ACTIONS.COLLECTION_ERROR, data: { message: 'collection does not exist' } });
+  const { uris: anyUris, remove, order, type } = params;
 
-  const editedCollection: Collection = makeSelectEditedCollectionForId(collectionId)(state);
-  const unpublishedCollection: Collection = makeSelectUnpublishedCollectionForId(collectionId)(state);
-  const publishedCollection: Collection = makeSelectPublishedCollectionForId(collectionId)(state); // needs to be published only
+  // -- sanitization --
+  // only permanent urls can be added to collections
+  let uris;
+
+  if (anyUris) {
+    uris = [];
+
+    anyUris.forEach(async (uri) => {
+      // related to selectBrokenUrlsForCollectionId
+      const isDeletingBrokenUris = typeof uri !== 'string';
+
+      // $FlowFixMe
+      if (isPermanentUrl(uri) || isDeletingBrokenUris) return uris.push(uri);
+
+      let url = selectPermanentUrlForUri(state, uri);
+
+      if (!url) {
+        const claim = await dispatch(doGetClaimFromUriResolve(url));
+        if (claim) url = claim.permanent_url;
+      }
+
+      // $FlowFixMe
+      return uris.push(url);
+    });
```

Comment:
- [ ] `await` in a `forEach` is undefined behavior. I have had issues recently trying to do something similar.
    - https://www.becomebetterprogrammer.com/javascript-foreach-async-await/#:~:text=Programmers%20use%20the%20forEach%20method,async%20and%20await%20with%20it.
- [ ] Also, what happens when the list is huge and every `url` happens to be undefined -- does it resolve and await one by one?